### PR TITLE
Proactively decrement scale set count during deletion operations 

### LINKED ISSUE #3031: Deletion requests and stale cache can cause scale set to go below configured minimum
In some cases, where there's two deletion candidates ready for deletion in close proximity to one another, and there's a deletion request in flight, the `getVMSSInfo` would return a stale VMSS size and a deletion request would still happen for the second.

We should forceRefresh the caches after sending a delete request so as to block further requests that could possibly go below the min size.

/area provider/azure
/kind bug

### PR DESCRIPTION
**Pull request by marwanad**
Fixes #3031 

This is to avoid stale cache reads which might deem a node deletable and cause the scale-down operation to below minimum node count.

/area provider/azure
/kind bug

### CODE REVIEW

File: cluster-autoscaler/cloudprovider/azure/azure_scale_set.go [hunk 1]
```diff
@@ -700,3 +709,13 @@ func (scaleSet *ScaleSet) invalidateInstanceCache() {
 	scaleSet.lastInstanceRefresh = time.Now().Add(-1 * vmssInstancesRefreshPeriod)
 	scaleSet.instanceMutex.Unlock()
 }
+
+func (scaleSet *ScaleSet) invalidateSizeCacheWithLock() {
+	scaleSet.sizeMutex.Lock()
+	scaleSet.lastSizeRefresh = time.Now().Add(-1 * scaleSet.sizeRefreshPeriod)
+	scaleSet.sizeMutex.Unlock()
+
+	scaleSetStatusCache.mutex.Lock()
```
Comment (marwanad) [inline cluster-autoscaler/cloudprovider/azure/azure_scale_set.go:718]: Pointing this diff out:

I don't think the initial intention was working as expected, from the comment:

```
// Invalidate the VMSS size cache in order to fetch the size from the API.
```

However, there's another safeguard in `getVMSSInfo` that would also read data from cache if `scaleSet.sizeRefreshPeriod` hasn't passed from the last `scaleSetStatusCache` refresh.

Comment (feiskyer) [reply cluster-autoscaler/cloudprovider/azure/azure_scale_set.go:718]: Right, this is not caught in our unit tests.  Could you help to add a unit test for the cache refreshing?

Comment (marwanad) [reply cluster-autoscaler/cloudprovider/azure/azure_scale_set.go:718]: yup, i'll look into adding one.

**Comment by feiskyer** [comment]
@marwanad could you also cherry-pick the fix to older branches?

**Pull request merged**